### PR TITLE
Added support for french layout bépo

### PR DESCRIPTION
**Pull request by shinze**
Added support for french layout Bépo http://bepo.fr/

### CODE REVIEW

File: keyboards/keebio/iris/keymaps/shinze/config.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef CONFIG_USER_H
```
Comment (fauxpark) [inline keyboards/keebio/iris/keymaps/shinze/config.h]:
```suggestion
#pragma once
```

File: keyboards/keebio/iris/keymaps/shinze/config.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+#define CONFIG_USER_H
```
Comment (fauxpark) [inline keyboards/keebio/iris/keymaps/shinze/config.h]:
```suggestion
```

File: keyboards/keebio/iris/keymaps/shinze/keymap.c [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  BEPO = SAFE_RANGE,
+  LOWER,
+  RAISE,
+  ADJUST,
+  ZO,
+  ZI
+};
+
+#define KC_ KC_TRNS
+#define _______ KC_TRNS
```
Comment (fauxpark) [inline keyboards/keebio/iris/keymaps/shinze/keymap.c]:
```suggestion
```

This alias is already defined in QMK so you don't need to define it here.

File: keyboards/keebio/iris/keymaps/shinze/keymap.c [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+
+  switch (keycode) {
+    case BEPO:
+      if (record->event.pressed) {
+        persistent_default_layer_set(1UL<<_BEPO);
+      }
+      return false;
+      break;
```
Comment (fauxpark) [inline keyboards/keebio/iris/keymaps/shinze/keymap.c]: `break` after `return` is dead code.

File: keyboards/keebio/iris/keymaps/shinze/rules.mk [hunk 5]
```diff
@@ -0,0 +1,6 @@
+RGBLIGHT_ENABLE = yes
+BACKLIGHT_ENABLE = yes
+
+ifndef QUANTUM_DIR
```
Comment (fauxpark) [inline keyboards/keebio/iris/keymaps/shinze/rules.mk]: Please remove this block - it's a holdover from the old Make system and no longer does anything.

File: keyboards/keebio/iris/keymaps/shinze/config.h [hunk 6]
```diff
@@ -0,0 +1,41 @@
+/*
+Copyright 2017 Danny Nguyen <danny@keeb.io>
```
Comment (fauxpark) [inline keyboards/keebio/iris/keymaps/shinze/config.h]: You might want to update the copyright here :)

File: keyboards/keebio/iris/keymaps/shinze/keymap.c [hunk 7]
```diff
@@ -0,0 +1,173 @@
+#include QMK_KEYBOARD_H
+
+#include "keymap_french_osx.h"
+#include "keymap_bepo.h"
+#include "sendstring_bepo.h"
+
+extern keymap_config_t keymap_config;
+
+#define _BEPO    0
```
Comment (fauxpark) [inline keyboards/keebio/iris/keymaps/shinze/keymap.c]: Non-consecutive layer numbers will generate extra empty layers in between, unnecessarily increasing flash usage. The better way to define layers is with an enum, eliminating the need to number them manually:

```c
enum layer_names {
    _BEPO,
    _LOWER,
    _RAISE,
    _ADJUST
};
```

File: keyboards/keebio/iris/keymaps/shinze/config.h [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef CONFIG_USER_H
+#define CONFIG_USER_H
+
+#include "config_common.h"
```
Comment (noroadsleft) [inline keyboards/keebio/iris/keymaps/shinze/config.h]: `keyboards/keebio/iris/config.h` includes this file; this line is unnecessary.

```suggestion
#include "config_common.h"
```

File: keyboards/keebio/iris/keymaps/shinze/config.h [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+// #define EE_HANDS
+
+#undef RGBLED_NUM
+#define RGBLIGHT_ANIMATIONS
+#define RGBLED_NUM 12
+#define RGBLIGHT_HUE_STEP 8
+#define RGBLIGHT_SAT_STEP 8
+#define RGBLIGHT_VAL_STEP 8
+
+#endif
```
Comment (noroadsleft) [inline keyboards/keebio/iris/keymaps/shinze/config.h]:
```suggestion
```

File: keyboards/keebio/iris/keymaps/shinze/keymap.c [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+  //,------+--------+--------+--------+--------+--------.                    ,-------+--------+---------+--------+--------+---------.
+    _______,_______ , _______ , _______ , KC_UP  ,_______  ,                  _______, _______ , _______ , _______, _______ ,BP_PERCENT,
+  //,------+--------+--------+--------+--------+--------+---------. ,--------,-------+--------+---------+--------+--------+---------.
+    _______, _______, _______ , KC_LEFT , KC_DOWN , KC_RIGHT,_______, _______,_______, _______ ,_______ , _______, _______ , _______,
+  //,------+--------+--------+--------+--------+--------+---------. ,--------,-------+--------+---------+--------+--------+---------.
+                                    BP_RALT, BP_RGUI, BP_LOWR,          SFT_T(BP_SPC), BP_ENT , BP_BSPC
+  //                               +--------+-------+---------.       ,--------------+--------+--------                                
+  ),
+
+  [_LOWER] = KC_KEYMAP(
```
Comment (drashna) [inline keyboards/keebio/iris/keymaps/shinze/keymap.c]: `KC_KEYMAP` no longer exists, and has been removed.

In general, we recommend against macros that prepend `KC_`, as they cause confusion.

But if you want to use it, you want:
```suggestion
  [_LOWER] = LAYOUT_kc
```

File: keyboards/keebio/iris/keymaps/shinze/keymap.c [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+    _______, _______, _______, _______, _______, _______,                     _______, _______, _______  , _______, _______ , _______,
+  //,------+--------+--------+--------+--------+--------+---------.  ,--------,-------+--------+---------+--------+--------+---------.
+    _______, _______, _______, _______, _______, _______, _______,    _______, _______, _______ ,_______ , _______, _______ , _______,
+  //,------+--------+--------+--------+--------+--------+---------.  ,--------,-------+--------+---------+--------+--------+---------.
+                                    _______, _______, _______,         _______, _______, _______
+  )
+
+};
+
+void persistent_default_layer_set(uint16_t default_layer) {
```
Comment (drashna) [inline keyboards/keebio/iris/keymaps/shinze/keymap.c]: This function is deprecated and should be removed,

File: keyboards/keebio/iris/keymaps/shinze/keymap.c [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+  eeconfig_update_default_layer(default_layer);
+  default_layer_set(default_layer);
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+
+  switch (keycode) {
+    case BEPO:
+      if (record->event.pressed) {
+        persistent_default_layer_set(1UL<<_BEPO);
```
Comment (drashna) [inline keyboards/keebio/iris/keymaps/shinze/keymap.c]:
```suggestion
        set_single_persistent_default_layer(_BEPO);
```

However, if you only have this default layer, you can remove the entire keycode, and the function for it.

File: keyboards/keebio/iris/keymaps/shinze/keymap.c (hunk 4 above)
Comment (drashna) [reply keyboards/keebio/iris/keymaps/shinze/keymap.c]: Yeah, but it's good practice, called or not. 

Also, it is better to remove the `return` here, than the `break`, as returning false may berak some detection.

Comment (fauxpark) [reply keyboards/keebio/iris/keymaps/shinze/keymap.c]: https://stackoverflow.com/questions/957038/break-in-a-case-with-return-and-for-default

> you will get warnings about unreachable code on some compilers.

Evidently not with QMK, but from the looks of it, doesn't seem like good practice to me.

**Comment by drashna** [comment]
Any updates on this?

**Comment by drashna** [comment]
Since there have been no updates or interaction in this PR, and it's been open for quite a while, we're closing this pull request.

If you'd like to update this PR, you can re-open it or resubmit it.

if you need any help, don't hesitate to ask us.

**Comment by drashna** [comment]
Since there have been no updates or interaction in this PR, and it's been open for quite a while, we're closing this pull request.

If you'd like to update this PR, you can re-open it or resubmit it.

if you need any help, don't hesitate to ask us.

**Pull request closed without merging**